### PR TITLE
[keymgr_dpe] Replace keymgr in EG - PR4 BootStageOwnerInt / Top dep. parameter

### PR DESCRIPTION
**Pull request by rroth-lowrisc**
# PR4 - BootStageOwnerInt / Top dep. parameter
This PR is the fourth in a series which will replace the [_keymgr_](https://opentitan.org/book/hw/ip/keymgr/index.html) with the [_kemgr_dpe_](https://opentitan.org/book/hw/ip/keymgr_dpe/index.html) in earlgrey. The replacement was approved by [this RFC](https://docs.google.com/document/d/1iF0EWyJkSEtRL9d057imLo4s6DWNrZ6Mpt0QKD8OMMc/).

## Merge-Dependencies

- #30617 
- #30615 

## Relevant Commits

Last two commits

## Description

The goal of this PR is twofold:

- Introduce the third bootstage required for the integration of the `keymgr_dpe` into earlgrey as approved by the RFC.
 - Make all necessary parameters top dependent which diverge between darjeeling and earlgrey (to avoid templating the IP)

After this PR is approved, both commits should be squashed together into a single commit.

### CODE REVIEW

File: hw/ip/keymgr_dpe/dv/env/seq_lib/keymgr_dpe_smoke_vseq.sv [hunk 1]
```diff
@@ -136,17 +136,17 @@ class keymgr_dpe_smoke_vseq extends keymgr_dpe_base_vseq;
     // set dst_slot to 1, to begin filling slot 1-3 again
     src_slot = 0;
     dst_slot = 1;
-    for (int iter = 0; iter<keymgr_dpe_pkg::DpeNumSlots; iter++) begin
+    for (int iter = 0; iter<keymgr_dpe_env_pkg::DvNumInstHwSlot; iter++) begin
```
Comment (gautschimi) [inline hw/ip/keymgr_dpe/dv/env/seq_lib/keymgr_dpe_smoke_vseq.sv]: `iter < keymgr_xx` (missing space)

File: hw/ip/keymgr_dpe/dv/env/seq_lib/keymgr_dpe_smoke_vseq.sv [hunk 2]
```diff
@@ -115,15 +115,15 @@ class keymgr_dpe_smoke_vseq extends keymgr_dpe_base_vseq;
     // only have boot_stage 1, which gives us enough boot_stages to advance and fill
     // key_slots 1-3 again.
     `uvm_info(`gfn, "Key Manager DPE smoke - erase test", UVM_LOW)
-    for (int iter = 1; iter<keymgr_dpe_pkg::DpeNumSlots; iter++) begin
+    for (int iter = 1; iter<keymgr_dpe_env_pkg::DvNumInstHwSlot; iter++) begin
```
Comment (gautschimi) [inline hw/ip/keymgr_dpe/dv/env/seq_lib/keymgr_dpe_smoke_vseq.sv]: missing space

File: hw/ip/keymgr_dpe/dv/env/keymgr_dpe_if.sv [hunk 3]
```diff
@@ -193,14 +192,14 @@ interface keymgr_dpe_if(input clk, input rst_n);
...
-    bit [NumRomDigestInputs-1:0] bad_rom_data = '0, bad_rom_valid = '0;
+    bit [DvNumRomDigestInputs-1:0] bad_rom_data = '0, bad_rom_valid = '0;
 
     repeat (num_invalid_input) begin
       randcase
         1: bad_keymgr_dpe_div = 1'b1;
         1: bad_otp_device_id = 1'b1;
-        1: bad_rom_data[$urandom % NumRomDigestInputs] = 1'b1;
-        1: bad_rom_valid[$urandom % NumRomDigestInputs] = 1'b1;
+        1: bad_rom_data[$urandom % DvNumRomDigestInputs] = 1'b1;
```
Comment (gautschimi) [inline hw/ip/keymgr_dpe/dv/env/keymgr_dpe_if.sv:201]: I don't understand what we are doing here. Why do we do random module num_digest?

Comment (gautschimi) [reply hw/ip/keymgr_dpe/dv/env/keymgr_dpe_if.sv:201]: Ah we are just picking a random entry, ok

File: hw/ip/keymgr_dpe/dv/env/keymgr_dpe_scoreboard.sv [hunk 4]
```diff
@@ -194,14 +210,26 @@ class keymgr_dpe_scoreboard extends cip_base_scoreboard #(
...
             if(boot_stage == keymgr_dpe_pkg::BootStageCreator) begin
-              `uvm_info(`gfn,
-              $sformatf({"process_kmac_data_req: boot_stage %0d is_err %0d",
-               "compare_boot_stage_0_data"},
-               boot_stage, is_err), UVM_LOW)
-              compare_boot_stage_0_data(
-                .exp_match(!is_err),
-                .byte_data_q(item.byte_data_q)
-              );
+              if (DvBootStages == 2) begin
```
Comment (gautschimi) [inline hw/ip/keymgr_dpe/dv/env/keymgr_dpe_scoreboard.sv]: This case differentiation could be handled inside the function `compare_boot_stage_0()`

instead of here with calls to two different functions which are doing similar things

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv [hunk 5]
```diff
@@ -460,27 +512,52 @@ module keymgr_dpe
...
+                                    rom_digest_vld &
+                                    creator_seed_vld;
+    assign adv_data_owner_int = '0;
+    assign adv_data_owner_int_valid = '0;
+  end else begin : gen_adv_matrix_for_three_boot_stages
+    assign adv_data_creator = DpeAdvDataWidth'({sw_binding,
+                                                device_id_i,
+                                                lc_keymgr_div_i,
+                                                rom_digests,
+                                                revision_seed});
```
Comment (gautschimi) [inline hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv:539]: Is there a reason why the order has changed? (position of `revision_seed`)

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv [hunk 6]
```diff
@@ -544,9 +621,8 @@ module keymgr_dpe
                                                   lc_keymgr_en[KeymgrDpeEnDebug]);
 
   // Only latch required signals to advance the bootstage during consumption.
-  logic is_creator_boot_stage, is_owner_boot_stage, is_owner_int_boot_stage;
+  logic is_creator_boot_stage, is_owner_boot_stage;
```
Comment (gautschimi) [inline hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv:624]: why was `is_owner_int_boot_stage` removed here? Wasn't it included in the last PR?

Comment (gautschimi) [reply hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv:624]: Ok I see, it was only used once and now the comparison is used instead of the generated signal. That's ok

**Review by gautschimi — COMMENTED** [review]
Thanks for this PR. It looks good from my side. Only a few minor comments/questions

File: hw/ip/keymgr_dpe/dv/env/keymgr_dpe_if.sv (hunk 3 above)
Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/dv/env/keymgr_dpe_if.sv:201]: Yes, in Earlgrey with only one ROM it doesn't make any difference - Darjeeling however supports two ROM's.

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv (hunk 5 above)
Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv:539]: **Earlgrey** and **Darjeeling** do not support the same ordering of seed material right now. Because `keymgr` and `keymgr_dpe` was developed for the corresponding top each, they reflect the ordering too.

In the future this could be changed so that both tops require the same ordering. However it would not change the RTL significantly as the `if ... else ...` could not be remove (due to the `creator_seed`).

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv (hunk 6 above)
Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv:624]: exactly

File: hw/ip/keymgr_dpe/dv/env/keymgr_dpe_scoreboard.sv (hunk 4 above)
Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/dv/env/keymgr_dpe_scoreboard.sv]: I moved it inside a common function, however the reuse of code is approximately similar.

**Comment by rroth-lowrisc** [comment]
~The last **[CI-Blocker]** commit is intentional to prevent the CI from running (at least partially) as discussed with @vogelpi . The reason is to avoid unnecessary runtime of the CI.~

~I will remove the commit as soon as I get review from the DV team.~

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv (hunk 5 above)
Comment (gautschimi) [reply hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv:539]: ok good, thanks

**Review by gautschimi — APPROVED** [review]

**Comment by gautschimi** [comment]
@rroth-lowrisc if you rebase this branch to master we don't have to reauthorize the changes. (because they are already in master)

**Comment by rroth-lowrisc** [comment]
> @rroth-lowrisc if you rebase this branch to master we don't have to reauthorize the changes. (because they are already in master)

I rebased this PR and #30617 to the latest master

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv [hunk 7]
```diff
@@ -541,18 +620,29 @@ module keymgr_dpe
...
   assign hw2reg.debug.invalid_health_state.de = adv_en & is_creator_boot_stage;
   assign hw2reg.debug.invalid_digest.de       = adv_en & is_creator_boot_stage;
 
-  // owner_seed is used when boot_stage is incremented from 1 (= OwnerInt) to 2 (= Owner).
+  // creator_seed is used when boot_stage is incremented from 1 (= OwnerInt) to
+  // 2 (= Owner) or from 0 (= Creator) to 1 (= OwnerInt) depending on the
+  // NumBootStages parameter
+  if (NumBootStages == 2) begin : gen_invalid_creator_seed_for_2_boot_stages
+    assign hw2reg.debug.invalid_creator_seed.de =
+        adv_en & is_owner_boot_stage;
```
Comment (andreaskurth) [inline hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv]:
```suggestion
  // creator_seed is consumed when boot_stage is incremented from 1 (= OwnerInt)
  // to 2 (= Owner) in the three-stage configuration, or from 0 (= Creator) to
  // 2 (= Owner) in the two-stage configuration (where OwnerInt is omitted).
  if (NumBootStages == 2) begin : gen_invalid_creator_seed_for_2_boot_stages
    assign hw2reg.debug.invalid_creator_seed.de =
        adv_en & is_creator_boot_stage;
```
Right?

**Comment by andreaskurth** [comment]
`theory_of_operation.md` could do with a few updates as well; suggestion:

```diff
diff --git a/hw/ip/keymgr_dpe/doc/theory_of_operation.md b/hw/ip/keymgr_dpe/doc/theory_of_operation.md
index fb4adae053..13c334f945 100644
--- a/hw/ip/keymgr_dpe/doc/theory_of_operation.md
+++ b/hw/ip/keymgr_dpe/doc/theory_of_operation.md
@@ -28,7 +28,7 @@ In order to address the relationships among keymgr_dpe slots and their stored DP
 
 ## Key Manager Slots
 
-Keymgr_dpe consists of `DpeNumSlots` slots, which is a generic parameter.
+Keymgr_dpe consists of `NumInstHwSlot` slots, which is a top-level parameter (instantiated up to `NumMaxHwSlot`).
 Each of these key manager slots can store a DPE context, i.e. all DICE-related information for a particular boot stage. That includes a secret key along with additional context information described below.
 The secret key size is fixed to 256-bit.
 
@@ -45,7 +45,8 @@ When a slot is active, `boot_stage` refers to its DPE context boot stage.
 For flexibility, `boot_stage` is a simple unsigned integer that is incremented from the parent's `boot_stage` during advance calls.
 Its actual mapping to boot stages such as ROM, BL0 or Kernel can be determined by SW.
 Hence, keymgr_dpe is oblivious to this mapping between counter values and the boot stages, though with an exception.
-The exception is that the first two stages are treated specially by RTL during KDF advance calls, as they consume other HW-backed inputs that come from keymgr_dpe’s peripheral inputs.
+The exception is that the initial boot stages are treated specially by RTL during KDF advance calls, as they consume other HW-backed inputs that come from keymgr_dpe’s peripheral inputs.
+How many stages consume such HW-backed inputs, and which inputs they consume, depends on the `NumBootStages` parameter (see [Advance](#advance)).
 From SW's point of view, they are still treated as any arbitrary DICE layer in that SW can provide further inputs through `SW_CDI_INPUT` CSR.
 
 A slot's `max_key_version` receives its value from `MAX_KEY_VERSION` register during a previous advance call that ends up populating this slot with DPE context.
@@ -132,11 +133,20 @@ If the OTP creator root key is not valid during the latching cycle, keymgr_dpe m
 
 Further advance calls use the key stored in the specified `CONTROL_SHADOWED.SLOT_SRC_SEL` slot (equally referred to as _parent_ or _source_ slot) , and the result of the derivation updates the slot specified by `CONTROL_SHADOWED.SLOT_DST_SEL`  (referred to as _destination_ or _child_ slot).
 Assuming that `key_policy`, `boot_stage` or `valid` bits of the parent context permit, the child secret is derived from the parent secret through a key derivation function during advance operation.
-`child_key = KDF(parent_key, message)`, where the message input might take few forms depending on the parent slot's `boot_stage` value.
-In particular:
-* If `boot_stage = 0` for the parent, then `message = SW_CDI_INPUT || hw_revision_seed || device_identifier || health_st_measurement || rom_descriptors || creator_seed`. See [KDF Details](#kdf-details) for more details on HW-backed inputs.
-* If `boot_stage = 1` for the parent, then `message = SW_CDI_INPUT || owner_seed`.
-* If `boot_stage > 1` for the parent, then `message = SW_CDI_INPUT`.
+`child_key = KDF(parent_key, message)`, where the message input takes different forms depending on the parent slot's `boot_stage` value.
+Which HW-backed inputs are consumed at each stage is controlled by the `NumBootStages` parameter, and in particular whether the `creator_seed` is consumed together with the other creator inputs (two-stage configuration) or by a dedicated intermediate owner stage (three-stage configuration).
+See [KDF Details](#kdf-details) for more details on HW-backed inputs.
+
+With `NumBootStages = 3`, the `creator_seed` is consumed by a dedicated `OwnerInt` (owner intermediate) stage:
+* If `boot_stage = 0` (Creator) for the parent, then `message = SW_CDI_INPUT || device_identifier || health_st_measurement || rom_descriptors || hw_revision_seed`.
+* If `boot_stage = 1` (OwnerInt) for the parent, then `message = SW_CDI_INPUT || creator_seed`.
+* If `boot_stage = 2` (Owner) for the parent, then `message = SW_CDI_INPUT || owner_seed`.
+* If `boot_stage > 2` for the parent, then `message = SW_CDI_INPUT`.
+
+With `NumBootStages = 2`, the `OwnerInt` stage is omitted; the `creator_seed` is consumed together with the other creator inputs, and an advance from `Creator` increments `boot_stage` directly to `Owner` (i.e. `boot_stage = 1` is not used):
+* If `boot_stage = 0` (Creator) for the parent, then `message = SW_CDI_INPUT || hw_revision_seed || device_identifier || health_st_measurement || rom_descriptors || creator_seed`.
+* If `boot_stage = 2` (Owner) for the parent, then `message = SW_CDI_INPUT || owner_seed`.
+* If `boot_stage > 2` for the parent, then `message = SW_CDI_INPUT`.
 
 At the end of a successful advance operation, the following updates are made for the slot selected by `SLOT_DST_SEL`:
 * `valid` bit is set to 1.
@@ -161,7 +171,7 @@ When there is no fault and the enable signal is active by life cycle controller,
   * If `retain_parent = true`, then the source and the destination slots are different.
   * If `retain_parent = true`, then the destination slot is not valid (i.e. `valid = 0`).
   * If `retain_parent = false`, then the source and the destination slots are the same.
-  * `boot_stage` of the source slot has not reached to the maximum value supported by HW (i.e. `boot_stage + 1 < DpeNumBootStages`.
+  * `boot_stage` of the source slot has not reached to the maximum value supported by HW (i.e. `boot_stage + 1 < NumBootStages`.
 
 
 ### Versioned Key Generation
@@ -212,7 +222,7 @@ During advance operations, KDF inputs are 0 padded to `AdvDataWidth` bits. Depen
 * `hw_revision_seed` is a 256-bit netlist constant.
 * `device_identifier` is a 256-bit non-secret device identifier. This value is received from peripheral OTP port.
 * `health_st_measurement` is a 128-bit domain separator (i.e. diversification constant) that depends on the life cycle stage. This value is received from peripheral LC port.
-* `rom_descriptors` are two hash values for ROM0, ROM1. Each digest is 256-bits. These values are received from their respective ROM controllers.
+* `rom_descriptors` are `NumRomDigestInputs` hash values received from the respective ROM controller(s). Each digest is 256-bits.
 * `creator_seed` is 256-bit creator secret received from the `SECRET2` OTP partition..
 * `owner_seed` is 256-bit owner secret received from the `SECRET3` OTP partition.
```

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv [hunk 8]
```diff
@@ -541,18 +620,29 @@ module keymgr_dpe
...
   assign hw2reg.debug.invalid_dev_id.de       = adv_en & is_creator_boot_stage;
   assign hw2reg.debug.invalid_health_state.de = adv_en & is_creator_boot_stage;
   assign hw2reg.debug.invalid_digest.de       = adv_en & is_creator_boot_stage;
 
-  // owner_seed is used when boot_stage is incremented from 1 (= OwnerInt) to 2 (= Owner).
+  // creator_seed is used when boot_stage is incremented from 1 (= OwnerInt) to
+  // 2 (= Owner) or from 0 (= Creator) to 1 (= OwnerInt) depending on the
+  // NumBootStages parameter
+  if (NumBootStages == 2) begin : gen_invalid_creator_seed_for_2_boot_stages
+    assign hw2reg.debug.invalid_creator_seed.de =
```
Comment (andreaskurth) [inline hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv:638]: The `invalid_creator_seed`/`invalid_owner_seed` register fields currently don't seem to be predicted/checked nor stimulated by DV. Do you have them on the testplan? If not, let's create an issue to track this (the issue should be created before this PR gets merged, but the resolution of the issue is not urgent).

File: hw/ip/keymgr_dpe/dv/keymgr_dpe_sim_cfg.hjson [hunk 9]
```diff
@@ -42,6 +42,11 @@
   sim_tops: ["keymgr_dpe_bind", "keymgr_dpe_cov_bind", "sec_cm_prim_count_bind",
              "sec_cm_prim_sparse_fsm_flop_bind", "sec_cm_prim_onehot_check_bind"]
 
+  // Default build options
+  build_opts: ["+define+DEF_DV_BOOT_STAGES=2",
+              "+define+DEF_DV_DPE_NUM_SLOT=8",
+              "+define+DEF_DV_NUM_ROM_DIGEST=2"]
+
```
Comment (andreaskurth) [inline hw/ip/keymgr_dpe/dv/keymgr_dpe_sim_cfg.hjson]: To my understanding, when the `earlgrey` and `darjeeling` sim configs import this "base" sim config, this list of build options gets concatenated rather than replaced. So these variant-specific build opts must **not** be part of `keymgr_dpe_sim_cfg.hjson`, otherwise the values given here end up *appended* to the intended values, and as some simulators use the *last* `+define`, the Earlgrey variant would silently compile with Darjeeling parameters, which would be wrong.

```suggestion
```

**Comment by rroth-lowrisc** [comment]
> `theory_of_operation.md` could do with a few updates as well; suggestion:
> 
> ```diff
> diff --git a/hw/ip/keymgr_dpe/doc/theory_of_operation.md b/hw/ip/keymgr_dpe/doc/theory_of_operation.md
> index fb4adae053..13c334f945 100644
> --- a/hw/ip/keymgr_dpe/doc/theory_of_operation.md
> +++ b/hw/ip/keymgr_dpe/doc/theory_of_operation.md
> @@ -28,7 +28,7 @@ In order to address the relationships among keymgr_dpe slots and their stored DP
>  
>  ## Key Manager Slots
>  
> -Keymgr_dpe consists of `DpeNumSlots` slots, which is a generic parameter.
> +Keymgr_dpe consists of `NumInstHwSlot` slots, which is a top-level parameter (instantiated up to `NumMaxHwSlot`).
>  Each of these key manager slots can store a DPE context, i.e. all DICE-related information for a particular boot stage. That includes a secret key along with additional context information described below.
>  The secret key size is fixed to 256-bit.
>  
> @@ -45,7 +45,8 @@ When a slot is active, `boot_stage` refers to its DPE context boot stage.
>  For flexibility, `boot_stage` is a simple unsigned integer that is incremented from the parent's `boot_stage` during advance calls.
>  Its actual mapping to boot stages such as ROM, BL0 or Kernel can be determined by SW.
>  Hence, keymgr_dpe is oblivious to this mapping between counter values and the boot stages, though with an exception.
> -The exception is that the first two stages are treated specially by RTL during KDF advance calls, as they consume other HW-backed inputs that come from keymgr_dpe’s peripheral inputs.
> +The exception is that the initial boot stages are treated specially by RTL during KDF advance calls, as they consume other HW-backed inputs that come from keymgr_dpe’s peripheral inputs.
> +How many stages consume such HW-backed inputs, and which inputs they consume, depends on the `NumBootStages` parameter (see [Advance](#advance)).
>  From SW's point of view, they are still treated as any arbitrary DICE layer in that SW can provide further inputs through `SW_CDI_INPUT` CSR.
>  
>  A slot's `max_key_version` receives its value from `MAX_KEY_VERSION` register during a previous advance call that ends up populating this slot with DPE context.
> @@ -132,11 +133,20 @@ If the OTP creator root key is not valid during the latching cycle, keymgr_dpe m
>  
>  Further advance calls use the key stored in the specified `CONTROL_SHADOWED.SLOT_SRC_SEL` slot (equally referred to as _parent_ or _source_ slot) , and the result of the derivation updates the slot specified by `CONTROL_SHADOWED.SLOT_DST_SEL`  (referred to as _destination_ or _child_ slot).
>  Assuming that `key_policy`, `boot_stage` or `valid` bits of the parent context permit, the child secret is derived from the parent secret through a key derivation function during advance operation.
> -`child_key = KDF(parent_key, message)`, where the message input might take few forms depending on the parent slot's `boot_stage` value.
> -In particular:
> -* If `boot_stage = 0` for the parent, then `message = SW_CDI_INPUT || hw_revision_seed || device_identifier || health_st_measurement || rom_descriptors || creator_seed`. See [KDF Details](#kdf-details) for more details on HW-backed inputs.
> -* If `boot_stage = 1` for the parent, then `message = SW_CDI_INPUT || owner_seed`.
> -* If `boot_stage > 1` for the parent, then `message = SW_CDI_INPUT`.
> +`child_key = KDF(parent_key, message)`, where the message input takes different forms depending on the parent slot's `boot_stage` value.
> +Which HW-backed inputs are consumed at each stage is controlled by the `NumBootStages` parameter, and in particular whether the `creator_seed` is consumed together with the other creator inputs (two-stage configuration) or by a dedicated intermediate owner stage (three-stage configuration).
> +See [KDF Details](#kdf-details) for more details on HW-backed inputs.
> +
> +With `NumBootStages = 3`, the `creator_seed` is consumed by a dedicated `OwnerInt` (owner intermediate) stage:
> +* If `boot_stage = 0` (Creator) for the parent, then `message = SW_CDI_INPUT || device_identifier || health_st_measurement || rom_descriptors || hw_revision_seed`.
> +* If `boot_stage = 1` (OwnerInt) for the parent, then `message = SW_CDI_INPUT || creator_seed`.
> +* If `boot_stage = 2` (Owner) for the parent, then `message = SW_CDI_INPUT || owner_seed`.
> +* If `boot_stage > 2` for the parent, then `message = SW_CDI_INPUT`.
> +
> +With `NumBootStages = 2`, the `OwnerInt` stage is omitted; the `creator_seed` is consumed together with the other creator inputs, and an advance from `Creator` increments `boot_stage` directly to `Owner` (i.e. `boot_stage = 1` is not used):
> +* If `boot_stage = 0` (Creator) for the parent, then `message = SW_CDI_INPUT || hw_revision_seed || device_identifier || health_st_measurement || rom_descriptors || creator_seed`.
> +* If `boot_stage = 2` (Owner) for the parent, then `message = SW_CDI_INPUT || owner_seed`.
> +* If `boot_stage > 2` for the parent, then `message = SW_CDI_INPUT`.
>  
>  At the end of a successful advance operation, the following updates are made for the slot selected by `SLOT_DST_SEL`:
>  * `valid` bit is set to 1.
> @@ -161,7 +171,7 @@ When there is no fault and the enable signal is active by life cycle controller,
>    * If `retain_parent = true`, then the source and the destination slots are different.
>    * If `retain_parent = true`, then the destination slot is not valid (i.e. `valid = 0`).
>    * If `retain_parent = false`, then the source and the destination slots are the same.
> -  * `boot_stage` of the source slot has not reached to the maximum value supported by HW (i.e. `boot_stage + 1 < DpeNumBootStages`.
> +  * `boot_stage` of the source slot has not reached to the maximum value supported by HW (i.e. `boot_stage + 1 < NumBootStages`.
>  
>  
>  ### Versioned Key Generation
> @@ -212,7 +222,7 @@ During advance operations, KDF inputs are 0 padded to `AdvDataWidth` bits. Depen
>  * `hw_revision_seed` is a 256-bit netlist constant.
>  * `device_identifier` is a 256-bit non-secret device identifier. This value is received from peripheral OTP port.
>  * `health_st_measurement` is a 128-bit domain separator (i.e. diversification constant) that depends on the life cycle stage. This value is received from peripheral LC port.
> -* `rom_descriptors` are two hash values for ROM0, ROM1. Each digest is 256-bits. These values are received from their respective ROM controllers.
> +* `rom_descriptors` are `NumRomDigestInputs` hash values received from the respective ROM controller(s). Each digest is 256-bits.
>  * `creator_seed` is 256-bit creator secret received from the `SECRET2` OTP partition..
>  * `owner_seed` is 256-bit owner secret received from the `SECRET3` OTP partition.
> ```

My original goal was to write a later on a documentation PR which introduces all the new features. But probably better to append a minimalistic version when changing the RTL.

File: hw/ip/keymgr_dpe/dv/keymgr_dpe_sim_cfg.hjson (hunk 9 above)
Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/dv/keymgr_dpe_sim_cfg.hjson]: I think as long as the simulators uses the last defined one it would work (never had a problem with failing smoketests) but I removed them nevertheless.

**Comment by andreaskurth** [comment]
> My original goal was to write a later on a documentation PR which introduces all the new features. But probably better to append a minimalistic version when changing the RTL.

Yes, let's please change everything that would become wrong or outdated in the documentation in the same PR as the RTL. Otherwise, we incur too much risk of not doing the required documentation updates and/or getting confused by doc-code mismatches in the meantime.

I agree that substantial extensions of the documentation to cover new features in greater detail can wait until follow-up PRs, but whenever we are planning to do that, we need an issue collecting a summary of the pending extensions to track the remaining work.

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv (hunk 8 above)
Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv:638]: ~I will add a test in the testplan, the PR is already open so the overhead is nearly zero~

Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/rtl/keymgr_dpe.sv:638]: ~Currently the testplan checks only if I have invalid hw input data e.g. all 0's or all 1's~

The current testplan has a test which verifies if the signals are valid, it only has to check the debug register too.

**Comment by rroth-lowrisc** [comment]
Thanks for the review @andreaskurth 
I squashed these two commits together to avoid introducing breaking commits.
AscentLint for _darjeeling_ is passing (`keymgr_dpe` is currently only included in the _darjeeling_ configuration)

**Review by andreaskurth — COMMENTED** [review]
LGTM, thanks @rroth-lowrisc!

Obviously, please rebase on `master` after #30617 has been merged and before merging this PR :-)

File: hw/ip/keymgr_dpe/dv/env/seq_lib/keymgr_dpe_base_vseq.sv [hunk 10]
```diff
@@ -100,7 +100,16 @@ class keymgr_dpe_base_vseq extends cip_base_vseq #(
       end
     end
 
-    `uvm_info(`gfn, "Initializating key manager", UVM_MEDIUM)
+    `uvm_info(`gfn, "Initializing keymgr dpe", UVM_MEDIUM)
+    `uvm_info(`gfn,
```
Comment (martin-velay) [inline hw/ip/keymgr_dpe/dv/env/seq_lib/keymgr_dpe_base_vseq.sv:104]: Nit: this comment is valid in multiple locations, your prints are spread over so many lines. I think you should better use more of the 100 width (which already ridiculously short for DV code).

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe_ctrl.sv [hunk 11]
```diff
@@ -12,7 +12,13 @@ module keymgr_dpe_ctrl
   import keymgr_dpe_pkg::*;
   import keymgr_dpe_reg_pkg::*;
 // TODO(#384): Bring back KmacEnMasking parameter
-(
+#(
+  // Number of instantiated HW slots
+  parameter int unsigned NumInstHwSlot = 4,
+  parameter int unsigned NumInstHwSlotWidth = 2,
```
Comment (martin-velay) [inline hw/ip/keymgr_dpe/rtl/keymgr_dpe_ctrl.sv:18]: I feel it's a bit odd to pass these 2 parameters as one is derived from the other. Additionally, if this block is used anywhere else there might be an issue with unmatching param. Why just dropping this parameter and doing a localparam calculation as done in `keymgr_dpe`?

File: hw/top_darjeeling/dv/env/seq_lib/chip_sw_keymgr_dpe_key_derivation_vseq.sv [hunk 12]
```diff
@@ -54,14 +54,18 @@ class chip_sw_keymgr_dpe_key_derivation_vseq extends chip_sw_base_vseq;
 
   localparam int KmacDigestBytes = kmac_pkg::AppDigestW / 8;
 
+  int num_hw_slots;
   bit lc_at_prod;
 
   virtual task dut_init(string reset_kind = "HARD");
+    string path = $sformatf("tb.dut.top_darjeeling.u_keymgr_dpe.NumInstHwSlot");
```
Comment (martin-velay) [inline hw/top_darjeeling/dv/env/seq_lib/chip_sw_keymgr_dpe_key_derivation_vseq.sv]: I am usure about that one, but I think you could directly refer to the hierarchical path without having to use the `uvm_hdl_read`. Something like?
`num_hw_slots = tb.dut.top_darjeeling.u_keymgr_dpe.NumInstHwSlot;`

File: hw/ip/keymgr_dpe/doc/theory_of_operation.md [hunk 13]
```diff
@@ -127,16 +127,25 @@ The only relevant registers (or register fields) during the first advance call a
 
 In particular, the destination slot for the UDS is chosen by SW, and there is no designated special slot for it.
 Moreover, since the destination slot for this first advance call has no parent, its `boot_stage` value is not incremented but initialized to `0`.
-This initial latching can only be done once until the next power cycle.
+This initial latching can be repeated with the _Load Root Key_ operation unless locked with the `LOAD_KEY_LOCK` register.
```
Comment (martin-velay) [inline hw/ip/keymgr_dpe/doc/theory_of_operation.md:130]: This param is mentioned, but I don't see anything in the actual code. Is it normal?

File: hw/ip/keymgr_dpe/dv/env/keymgr_dpe_scoreboard.sv [hunk 14]
```diff
@@ -166,35 +199,55 @@ class keymgr_dpe_scoreboard extends cip_base_scoreboard #(
...
             end else begin
               `uvm_info(`gfn,
-              $sformatf({"process_kmac_data_req: boot_stage %0d is_err %0d",
-               "compare_boot_stage_gte_2_data"},
-               boot_stage, is_err), UVM_LOW)
+                        $sformatf({"process_kmac_data_req: boot_stage %0d ",
+                                   "is_err %0d compare_boot_stage_2_data"},
+                                   boot_stage, is_err),
+                        UVM_LOW)
                compare_boot_stage_gte_2_data(
```
Comment (martin-velay) [inline hw/ip/keymgr_dpe/dv/env/keymgr_dpe_scoreboard.sv]: Nit: this function here should probably be renamed as it's not really a stage 2 anymore

File: hw/ip/keymgr_dpe/doc/theory_of_operation.md (hunk 13 above)
Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/doc/theory_of_operation.md:130]: No, the original documentation was not properly updated when the feature `LOAD_KEY_LOCK` was introduced. When I read the `theory_of_operation` when reviewing @andreaskurth comment I saw the error and I fixed it.

File: hw/top_darjeeling/dv/env/seq_lib/chip_sw_keymgr_dpe_key_derivation_vseq.sv (hunk 12 above)
Comment (rroth-lowrisc) [reply hw/top_darjeeling/dv/env/seq_lib/chip_sw_keymgr_dpe_key_derivation_vseq.sv]: Unfortunatly this doesn't work. The simulation fails with:

```
src/lowrisc_dv_top_darjeeling_chip_env_0.1/seq_lib/chip_sw_keymgr_dpe_key_derivation_vseq.sv, 68
chip_env_pkg, "tb.dut.top_darjeeling.u_keymgr_dpe.NumInstHwSlot"
    Hierarchical reference from package 'chip_env_pkg' into module 'tb' is not allowed.
```

File: hw/ip/keymgr_dpe/dv/env/keymgr_dpe_scoreboard.sv (hunk 14 above)
Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/dv/env/keymgr_dpe_scoreboard.sv]: I renamed the function to `compare_boot_stage_runtime_data`. The corresponding _bootstage_ is called `BootStageRuntime` which makes far more sense.

File: hw/ip/keymgr_dpe/rtl/keymgr_dpe_ctrl.sv (hunk 11 above)
Comment (rroth-lowrisc) [reply hw/ip/keymgr_dpe/rtl/keymgr_dpe_ctrl.sv:18]: My goal was to have a "single source of truth" for both parameters throughout the `keymgr_dpe` in the case the relationship is changed between the two (something I would never expect to happen).

In addition I already need the size in the port list (signal `slot_src_sel_i` and `slot_dst_sel_i`).

**Comment by rroth-lowrisc** [comment]
@martin-velay I implemented your review
Could you check the file `chip_sw_keymgr_dpe_key_derivation_vseq.sv` again?
I had there some additional (minor) changes since you reviewed the file.

**Review by martin-velay — APPROVED** [review]
It looks good, I haven't got time to review it deeply though and the Darjeeling top vseq would need a complete rework at some point anyway.

**Review by andreaskurth — APPROVED** [review]
Thanks @rroth-lowrisc!
